cache/src/main/java/net/runelite/cache/MapImageDumper.java

```diff
@@ -639,6 +633,661 @@ else if (shape == 1)
 		}
 	}
 
+	private void setPixels(int[][] pixels, int x, int y, int value)
+	{
+		if (x >= 0 && y >= 0 && x < pixels.length && y < pixels[x].length)
+		{
+			pixels[x][y] = value;
+		}
+		else
+		{
+			System.err.println("Ground drawing out of bounds! " + x + ", " + y);
+		}
+	}
+
+	private int convertTileRotation(int rotation, int shape)
+	{
+		if (shape == 9)
+		{
+			rotation = rotation + 1 & 3;
+		}
+
+		if (shape == 10)
+		{
+			rotation = rotation + 3 & 3;
+		}
+
+		if (shape == 11)
+		{
+			rotation = rotation + 3 & 3;
+		}
+
+		return rotation;
+	}
+
+	int convertTileShape(int shape)
+	{
+		return shape != 9 && shape != 10 ? (shape == 11 ? 8 : shape) : 1;
+	}
+
+	private void generateTileShapes()
+	{
+		TILE_SHAPE_2D = new byte[8][4][];
+		genTileShapes1();
+		genTileShapes2();
+		genTileShapes3();
+		genTileShapes4();
+		genTileShapes5();
+		genTileShapes6();
+		genTileShapes7();
+		genTileShapes8();
```

Comment:
These methods all look pretty similar. Is there a way to combine them based on the `TILE_SHAPE_2D` index?

Reply:
Jagex has the logic broken out the same way, it's used to draw the smooth edges of overlay shapes (think paths and edges of islands). Looks to be taken right out of the client but stored into that result array so it isn't computed over and over.

In the reaally old mapview applet (2004-2008) it looks like
```
drawMap(...) {
    if overlay {
        if shape {
            drawSmoothEdges(...);
        }
    }
}

drawSmoothEdges(...) {
    // inline convertTileRotation() equivalent
    // followed by inline genTileShapes() that draw directly to the image buffer
}
```